### ISSUE TITLE
Sorting view results

### ISSUE THREAD

**Issue by jamesrkg** [Quantipy/quantipy#15]
We need a way to sort the results of views.

For example, given the following view:

```python
Question            gender  
Values              @
Question    Values  
q1          1       7.201447
            2       9.64925
            3       55.662665
            4       72.663656
            5       4.704566
            6       11.496034
            7       21.69883
            8       3.178551
            9       0.099151
            96      2.199429
            98      2.497416
            99      8.949005
```

We need a way to sort the index, including the ability to exclude or fix some values (for example, in this case the values 96, 98 and 99 which represent some form of non-response like "Don't know", "Prefer not to say", "None of these" and so on.

```python
Question            gender          
Values              @
Question    Values          
q1          4       72.663656
            3       55.662665
            7       21.69883
            6       11.496034
            2       9.64925
            1       7.201447
            5       4.704566
            8       3.178551
            9       0.099151
            96      2.199429
            98      2.497416
            99      8.949005
```

We also need a way to apply the sorted order of on view onto other views using the same x-variable. For example in this case you would probably want all other links using ``q1`` in the x-position to be similarly sorted.

```python
Question            gender          
Values              1           2
Question    Values          
q1          4       36.38664    36.277016
            3       27.682186   27.980479
            7       11.310729   10.388101
            6       5.035425    6.460609
            2       5.187247    4.462003
            1       3.669028    3.532419
            5       2.403846    2.30072
            8       1.644737    1.533814
            9       0.075911    0.02324
            96      1.037449    1.16198
            98      0.986842    1.510574
            99      4.57996     4.369045
```

There are a few possible approaches here. A quick solution (the focus of this ticket) will be to take a stack and sort links permanently (until re-sorted). A more sophisticated approach would be to decide on sorting when calling a build (this will not be the approach at this time).

**Comment by jamesrkg** [comment]
I'll implement this using column meta tags that are picked up by paint_dataframe. This means that cascade behaviour across a chain, for example, won't be needed.

**Comment by jamesrkg** [comment]
I've implemented this in a new meta object called ``rules``, using an option called ``sortx``.

The ``rules`` object is a dict with ``x`` and ``y`` keys for axis-specific rules application. Internally ``sortx`` results in a call to a function of the same name. The contents of the ``slicex`` object becomes that function call's kwargs.

```python
def sortx(df, sort_col='All', ascending=False, fixed=None):
    """
    Sort the index of df on a column, keeping margins and fixing values.
    
    This function sorts df, which is assumed to be a Quantipy-style
    view result with appropriate index/column structure, using
    a given column, while maintaining the position of margins if
    they exist, and also optionally fixing certain values at the
    bottom of the result without sorting them. Note that nested
    variable view results are not yet supported.
    
    Parameters
    ----------
    df : pandas.DataFrame
        The Quantipy-style view result to be sorted
    sort_col : str or int, default='All'
        The column (on the innermost level of the column's
        MultiIndex) on which to sort.
    ascending : bool, default=False
        Sort ascending vs. descending. Default descending for
        easier application to MR use cases.
    fixed : list-like, default=None
        A list of index values that should appear underneath
        the sorted index values.
    
    Returns
    -------
    df : pandas.DataFrame
        The sorted df. 
    """
```

Given the following crosstab:

```python
print crosstab(meta, data, 'income', 'income_cb')

Question                income_cb                                                                        
Values                        All    1    2    3    4    5    6   7   8   9   10   11   12  901  902  903
Question    Values                                                                                       
income      All              2078  425  204  187  146  119  103  89  68  59  226  101  351  425  848  353
            1                 425  425    0    0    0    0    0   0   0   0    0    0    0  425    0    0
            2                 204    0  204    0    0    0    0   0   0   0    0    0    0    0  204    0
            3                 187    0    0  187    0    0    0   0   0   0    0    0    0    0  187    0
            4                 146    0    0    0  146    0    0   0   0   0    0    0    0    0  146    0
            5                 119    0    0    0    0  119    0   0   0   0    0    0    0    0  119    0
            6                 103    0    0    0    0    0  103   0   0   0    0    0    0    0  103    0
            7                  89    0    0    0    0    0    0  89   0   0    0    0    0    0   89    0
            8                  68    0    0    0    0    0    0   0  68   0    0    0    0    0    0   68
            9                  59    0    0    0    0    0    0   0   0  59    0    0    0    0    0   59
            10                226    0    0    0    0    0    0   0   0   0  226    0    0    0    0  226
            11                101    0    0    0    0    0    0   0   0   0    0  101    0    0    0    0
            12                351    0    0    0    0    0    0   0   0   0    0    0  351    0    0    0
```

Set up the ``sortx`` ``rules``:

```python
meta['columns']['AMA_qincome']['rules'] = {
    'x': {'sortx': {'fixed': [1, 12, 10, 11]}}
}

meta['columns']['AMA_qincome_cb']['rules'] = {
    'y': {'sortx': {'fixed': [901, 902, 903], 'ascending': True}}
}
```

Apply the ``rules`` using crosstab:

```python
print crosstab(meta, data, 'income', 'income_cb', rules=True)

Question                income_cb                                                                        
Values                        All   9   8   7   11    6    5    4    3    2   10   12    1  901  902  903
Question    Values                                                                                       
income      All              2078  59  68  89  101  103  119  146  187  204  226  351  425  425  848  353
            2                 204   0   0   0    0    0    0    0    0  204    0    0    0    0  204    0
            3                 187   0   0   0    0    0    0    0  187    0    0    0    0    0  187    0
            4                 146   0   0   0    0    0    0  146    0    0    0    0    0    0  146    0
            5                 119   0   0   0    0    0  119    0    0    0    0    0    0    0  119    0
            6                 103   0   0   0    0  103    0    0    0    0    0    0    0    0  103    0
            7                  89   0   0  89    0    0    0    0    0    0    0    0    0    0   89    0
            8                  68   0  68   0    0    0    0    0    0    0    0    0    0    0    0   68
            9                  59  59   0   0    0    0    0    0    0    0    0    0    0    0    0   59
            1                 425   0   0   0    0    0    0    0    0    0    0    0  425  425    0    0
            12                351   0   0   0    0    0    0    0    0    0    0  351    0    0    0    0
            10                226   0   0   0    0    0    0    0    0    0  226    0    0    0    0  226
            11                101   0   0   0  101    0    0    0    0    0    0    0    0    0    0    0
```

**Comment by jamesrkg** [comment]
The ``rules`` object also supports ``slicex`` (#19) and ``dropx`` (#18), all of which may be used in any combination.

**Comment by jamesrkg** [comment]
When drawing chains out of a stack, rules can be applied to what is extracted in the same way:

```python
chains = stack.get_chain(..., rules=True)
```

``rules`` will currently only be applied to chain items if the parameter ``post_process`` is also ``True`` (which it is by default).

**Comment by jamesrkg** [comment]
Resolved by #40
